# 02-BOW

Now that we master the preprocessing, let's make our first Bag Of Words (BOW).

We will reuse our dataset of Coldplay songs to make a BOW.

As usual, the first step is to import some libraries. So import *nltk* as well as all the libraries you will need.

In [3]:
# Import NLTK and all the needed libraries
import nltk
import numpy as np
import pandas as pd
from sklearn.feature_extraction.text import CountVectorizer

Load now the dataset in *coldplay.csv* using pandas.

In [4]:
# TODO: Load the dataset in coldplay.csv
import pandas as pd

df = pd.read_csv('coldplay.csv')

print(df.head())



     Artist                           Song  \
0  Coldplay                 Another's Arms   
1  Coldplay                Bigger Stronger   
2  Coldplay                       Daylight   
3  Coldplay                       Everglow   
4  Coldplay  Every Teardrop Is A Waterfall   

                                                Link  \
0            /c/coldplay/anothers+arms_21079526.html   
1          /c/coldplay/bigger+stronger_20032648.html   
2                 /c/coldplay/daylight_20032625.html   
3                 /c/coldplay/everglow_21104546.html   
4  /c/coldplay/every+teardrop+is+a+waterfall_2091...   

                                              Lyrics  
0  Late night watching tv  \nUsed to be you here ...  
1  I want to be bigger stronger drive a faster ca...  
2  To my surprise, and my delight  \nI saw sunris...  
3  Oh, they say people come  \nThey say people go...  
4  I turn the music up, I got my records on  \nI ...  


You already know this dataset, but you can check it again if you want to refresh your memory.

In [7]:
# TODO: Explore the data
df.info()


<class 'pandas.core.frame.DataFrame'>
RangeIndex: 120 entries, 0 to 119
Data columns (total 4 columns):
 #   Column  Non-Null Count  Dtype 
---  ------  --------------  ----- 
 0   Artist  120 non-null    object
 1   Song    120 non-null    object
 2   Link    120 non-null    object
 3   Lyrics  120 non-null    object
dtypes: object(4)
memory usage: 3.9+ KB


Now using the *CountVectorizer* of scikit-learn, make a BOW of all the lyrics of Coldplay, and print the result.

In [20]:
# TODO: Compute a BOW
vectorizer = CountVectorizer(stop_words='english', analyzer='word')
bow = vectorizer.fit_transform(df['Lyrics']).toarray()
print(bow.shape)


(3, 20)


Now that we have the BOW matrix, we would like to have a new dataframe having the BOW for each song, and as columns the corresponding words (just as we did in the lecture at the end).

So that at the end we would end up with a dataframe containing something like the following (120 raws for 120 songs, and as many columns as words):

| | ah | adventure | ... | yeah
|---|---|---|---|---|
| 0 | 0 | 1 | ... | 4 |
| 1 | 8 | 0 | ... | 2 |
|...|...|...|...|...|
| 119 | 5 | 0 | ... | 8 |

In [21]:
# TODO: Create a new dataframe containing the BOW outputs and the corresponding words as columns. And print it
import pandas as pd
from sklearn.feature_extraction.text import CountVectorizer

vectorizer = CountVectorizer(stop_words='english', analyzer='word')

bow = vectorizer.fit_transform(df['Lyrics']).toarray()

words = vectorizer.get_feature_names_out()

bow_df = pd.DataFrame(bow, columns=words)

# Print the new DataFrame
print(bow_df)





   best  don  gave  look  morning  need  rise  rule  seas  shine  sleep  \
0     0    0     0     2        0     0     0     0     0      1      0   
1     0    0     1     0        1     0     1     1     1      0      1   
2     1    1     0     0        0     1     0     0     0      0      0   

   stars  succeed  try  used  want  word  world  yeah  yellow  
0      1        0    0     0     0     0      0     1       1  
1      0        0    0     1     0     1      1     0       0  
2      0        1    1     0     1     0      0     0       0  


Well as you see we're still having some issue, we have some tokens that are not words, like '10' or '2000'.

To get rid of that, we could use directly regular expressions within the function. Another solution would be to make preprocessing before using the function *CountVectorizer*.

For the moment, we won't pay attention to this issue. But if you are curious and have time, you can find on google how to remove those words using the *CountVectorizer*.

Now we would like to see what are the most used words by Coldplay.

In [22]:
sum_bow = bow_df.sum()
sum_bow.idxmax()

'look'

So what is the most used word? Are you surprised?

Now make a sort in order to show the 10 most used.

In [25]:
# TODO: print the 10 most used word by Coldplay
sum_bow.sort_values(ascending=False)[:10]


look       2
best       1
stars      1
yeah       1
world      1
word       1
want       1
used       1
try        1
succeed    1
dtype: int64

Here it is! You know the Coldplay lyrics more than the singers now!